### DISCUSSION TITLE
Integrate with external image codec

### DISCUSSION THREAD

**Discussion by dogtopus** [image-rs/image#1810] [Q&A]
What is the best approach to integrate the image crate with an external codec? I'm looking for specifically what intermediate steps to do to efficiently and safely load an image into a say `DynamicImage` type, as well as saving a `DynamicImage` type as the target format.

The high-level API seems to be hardcoded to only use the built-in codecs via an enum and not taking a codec that implements the `ImageDecoder`/`ImageEncoder` trait so that's off the table.

**Comment by fintelia** [comment]
I'm not sure that there's one single best option. For loading, you can implement `ImageDecoder` and then call [`DynamicImage::from_decoder`](https://docs.rs/image/latest/image/enum.DynamicImage.html#method.from_decoder). Or just directly write a method that takes a reader and returns a `DynamicImage`. We don't have a similar convenience method for encoding, but given a dynamic image the `DynamicImage::color` and `DynamicImage::as_bytes` methods should give you the information you need to implement your encoder.
